### PR TITLE
feat: foxy charts

### LINKED ISSUE #1317: FOXy balance charts
please don't spank me i'm a bounty hunter

image

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

* **do not merge until https://github.com/shapeshift/lib/pull/497 is merged**

* moves tx history in txHistory slice under txs namespace
* adds rebases namespace to txHistory slice
* adds getFoxyRebaseHistoryByAccountId api function
* fetches foxy rebase history on app load if flagged on
* considers rebase events in balance charts and makes them work

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1317

## Risk

this changes where we store tx history in the store, i've tested and it's functionally working

## Testing

* use a wallet with a foxy balance and rebase history - ask @toshiSat for a demonstration

## Screenshots (if applicable)

n/a - charts should go to zero back in time for foxy with a rebase history

### CODE REVIEW

File: src/state/slices/txHistorySlice/txHistorySlice.ts [hunk 1]
```diff
@@ -171,6 +257,40 @@ export const txHistoryApi = createApi({
...
+          const data = `getFoxyRebaseHistoryByAccountId: ChainAdapterManager does not support ${ChainTypes.Ethereum}`
+          const status = 418
+          const error = { data, status }
+          return { error }
+        }
+        const adapter = await adapters.byChainId(chainId)
+        const providerUrl = getConfig().REACT_APP_ETHEREUM_NODE_URL
+        const foxyArgs = { adapter, foxyAddresses, providerUrl }
+        const foxyApi = new FoxyApi(foxyArgs)
+        const tokenContractAddress = foxyAddresses[0].foxy.toLowerCase()
```
Comment (0xdef1cafe) [inline src/state/slices/txHistorySlice/txHistorySlice.ts:283]: @toshiSat do you wanna make this safer?

File: src/context/TransactionsProvider/TransactionsProvider.tsx [hunk 2]
```diff
@@ -90,16 +94,31 @@ export const TransactionsProvider = ({ children }: TransactionsProviderProps): J
...
+           * because we need it for the dashboard balance chart
+           *
+           * if you're reading this and are about to add another rebase token here,
+           * stop, and make a getRebaseHistoryByAccountId that takes
+           * an accountId and assetId[] in the txHistoryApi
+           */
+          if (getConfig().REACT_APP_FEATURE_FOXY_INVESTOR) {
+            // fetch all rebase history for FOXy
+            // you see what i did here?
+            if (!portfolioAssetIds.join('').includes(foxyAddresses[0].foxy.toLowerCase())) return
```
Comment (0xdef1cafe) [inline src/context/TransactionsProvider/TransactionsProvider.tsx:119]: @toshiSat wanna make this safer too?

File: src/state/slices/txHistorySlice/txHistorySlice.ts [hunk 3]
```diff
@@ -57,12 +59,33 @@ export type TxIdByAccountId = {
...
-  ids: TxId[]
-  status: TxHistoryStatus
+  txs: {
+    byId: TxHistoryById
+    byAssetId: TxIdByAssetId
+    byAccountId: TxIdByAccountId
+    ids: TxId[]
+    status: TxHistoryStatus
+  }
+  rebases: {
```
Comment (0xApotheosis) [inline src/state/slices/txHistorySlice/txHistorySlice.ts:83]: Can we type `txs` and `rebases`?

File: src/state/slices/txHistorySlice/txHistorySlice.ts [hunk 4]
```diff
@@ -142,23 +174,77 @@ const updateOrInsert = (txHistory: TxHistory, tx: Tx, accountSpecifier: AccountS
   // get applied to state when it goes out of scope
 }
 
+type UpdateOrInsertRebase = (txState: TxHistory, data: RebaseHistoryPayload['payload']) => void
+
+const updateOrInsertRebase: UpdateOrInsertRebase = (txState, payload) => {
```
Comment (0xApotheosis) [inline src/state/slices/txHistorySlice/txHistorySlice.ts:179]: Just a thought, but we could write this logic in a more pure way (note, not tested):

```ts
const updateOrInsertRebase: UpdateOrInsertRebase = (txState, payload) => {
  const { accountId, assetId } = payload
  let { rebases } = txState
  rebases = (() => {
    const byId: RebaseById = payload.data.map(rebase => {
      const rebaseId = makeRebaseId({ accountId, assetId, rebase })
      return { [rebaseId]: rebase }
    })

    const ids = payload.data.reduce((acc, rebase) => {
      const rebaseId = makeRebaseId({ accountId, assetId, rebase })
      const isNew = !txState.rebases.byId[rebaseId]
      if (isNew) {
        const orderedRebases = orderBy(rebases.byId, 'blockTime', ['desc'])
        const index = orderedRebases.findIndex(
          rebase => makeRebaseId({ accountId, assetId, rebase }) === rebaseId
        )
        return { ...acc, ids: slice(rebases.ids, index).concat(rebaseId) }
      } else {
        return acc
      }
    })

    type Rebase = {
      byAssetId: RebaseByAssetId
      byAccountId: RebaseByAccountId
      ids: RebaseId[]
      byId: RebaseById
    }

    const byAssetId: RebaseByAssetId = payload.data.reduce((acc: Rebase, rebase: Rebase) => {
      acc.byAssetId[assetId] = addToIndex(
        rebases.ids,
        rebases.byAssetId[assetId],
        makeRebaseId({ accountId, assetId, rebase })
      )
      return acc
    }).byAssetId

    const byAccountId: RebaseByAccountId = payload.data.reduce((acc: Rebase, rebase: Rebase) => {
      acc.byAccountId[assetId] = addToIndex(
        rebases.ids,
        rebases.byAccountId[accountId],
        makeRebaseId({ accountId, assetId, rebase })
      )
      return acc
    }).byAccountId

    return {
      ...rebases,
      byId,
      ids,
      byAssetId,
      byAccountId
    }
  })()
}
```

File: src/state/slices/txHistorySlice/txHistorySlice.ts [hunk 5]
```diff
@@ -171,6 +257,40 @@ export const txHistoryApi = createApi({
...
+        const chainId = caip2.toCAIP2({ chain, network })
+        const entries = Object.entries(accountSpecifierMap)[0]
+        const [accountChainId, userAddress] = entries
+        const accountId = entries.join('/')
+        // [] is a valid return type and won't upsert anything
+        if (chainId !== accountChainId) return { data: [] }
+        const adapters = getChainAdapters()
+        if (!adapters.getSupportedChains().includes(ChainTypes.Ethereum)) {
+          const data = `getFoxyRebaseHistoryByAccountId: ChainAdapterManager does not support ${ChainTypes.Ethereum}`
+          const status = 418
```
Comment (0xApotheosis) [inline src/state/slices/txHistorySlice/txHistorySlice.ts:275]: Can we remove this magic number?

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Nice, some thoughts.

File: src/state/slices/txHistorySlice/txHistorySlice.ts (hunk 5 above)
Comment (0xdef1cafe) [reply src/state/slices/txHistorySlice/txHistorySlice.ts:277]: siiiiigh i'll use 400 😝

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 6]
```diff
@@ -121,31 +125,43 @@ export const makeBuckets: MakeBuckets = args => {
...
       console.error(
-        `bucketTxs: tx outside buckets: ${
-          tx.txid
-        }, start: ${start.valueOf()}, end: ${end.valueOf()}, meta: ${meta}`
+        `bucketTxs: event outside buckets: ${event}, start: ${start.valueOf()}, end: ${end.valueOf()}, meta: ${meta}`
       )
       return acc
     }
+
+    const isTx = (event: Tx | RebaseHistory): event is Tx => Boolean((event as Tx)?.txid ?? '')
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:155]: fixed

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 7]
```diff
@@ -121,31 +125,43 @@ export const makeBuckets: MakeBuckets = args => {
...
     // add to the correct bucket
-    acc[bucketIndex].txs.push(tx)
+    isTx(event)
+      ? acc[bucketIndex].txs.push(event as Tx)
+      : acc[bucketIndex].rebases.push(event as RebaseHistory)
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:159]: fixed

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 8]
```diff
@@ -242,6 +258,13 @@ export const calculateBucketPrices: CalculateBucketPrices = args => {
       })
     })
 
+    rebases.forEach(rebase => {
+      const { assetId, balanceDiff } = rebase
+      if (!assetIds.includes(assetId)) return
+      // UP ONLY
+      bucket.balance.crypto[assetId] = bnOrZero(bucket.balance.crypto[assetId]).minus(balanceDiff)
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:265]: the mutation is actually intentional in this particular hook - we can get the whole computation down to nearly a single frame (16ms) in most cases

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 9]
```diff
@@ -121,31 +125,43 @@ export const makeBuckets: MakeBuckets = args => {
...
     const { duration, unit } = meta
     // the number of time units from start of chart to this tx
-    let bucketIndex = Math.floor(txDayjs.diff(start, unit as dayjs.OpUnitType) / duration)
+    let bucketIndex = Math.floor(eventDayJs.diff(start, unit as dayjs.OpUnitType) / duration)
     if (bucketIndex < 0 || bucketIndex > buckets.length - 1) {
       console.error(
-        `bucketTxs: tx outside buckets: ${
-          tx.txid
-        }, start: ${start.valueOf()}, end: ${end.valueOf()}, meta: ${meta}`
+        `bucketTxs: event outside buckets: ${event}, start: ${start.valueOf()}, end: ${end.valueOf()}, meta: ${meta}`
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:150]: yeah - as we have to narrow the type to access particular fields, and the only intersection between types is `blockTime`. really - this was originally here for debugging when building the charts the first pass around, to trigger it you'd need a wallet with tx history > 5 years

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 10]
```diff
@@ -242,6 +258,13 @@ export const calculateBucketPrices: CalculateBucketPrices = args => {
       })
     })
 
+    rebases.forEach(rebase => {
+      const { assetId, balanceDiff } = rebase
+      if (!assetIds.includes(assetId)) return
+      // UP ONLY
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:262]: https://uponly.tv/ let me have my fun ser

File: src/state/slices/txHistorySlice/txHistorySlice.ts (hunk 4 above)
Comment (0xdef1cafe) [reply src/state/slices/txHistorySlice/txHistorySlice.ts:183]: i really like this pattern, and am going to steal it along with try/catch IIFE style, but if i update this i want to update the txHistory upsert too, and don't want to fix regressions here. i know the slice logic changes looks suspect at first glance (i know you said untested here)

File: src/hooks/useBalanceChartData/useBalanceChartData.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (0xApotheosis) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts]: I learn much about the crypto sphere by reviewing your PRs 😂

File: src/context/TransactionsProvider/TransactionsProvider.tsx [hunk 11]
```diff
@@ -90,16 +94,31 @@ export const TransactionsProvider = ({ children }: TransactionsProviderProps): J
...
+          const forceRefetch = true
+          const options = { forceRefetch }
+          dispatch(getAllTxHistory.initiate({ accountSpecifierMap }, options))
+
+          /**
+           * foxy rebase history is most closely linked to tranasctions.
+           * unfortunately, we have to call this for a specific asset here
+           * because we need it for the dashboard balance chart
+           *
+           * if you're reading this and are about to add another rebase token here,
```
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/context/TransactionsProvider/TransactionsProvider.tsx:111]: ok ok ur right

File: src/state/slices/txHistorySlice/txHistorySlice.ts (hunk 4 above)
Comment (0xApotheosis) [reply src/state/slices/txHistorySlice/txHistorySlice.ts:183]: It's also slightly less performant, so not always the best choice - but generally, I prefer it as it makes spotting bugs and understanding what's going on way easier.

Yeah, I'm not sure the reducer quite works and didn't wanna waste time on it, but it's something like that.

**Review by DaoDev44 — APPROVED** [review]

File: src/context/TransactionsProvider/TransactionsProvider.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (0xdef1cafe) [reply src/context/TransactionsProvider/TransactionsProvider.tsx:109]: ok, i had a bash at this - the `queryFn` inside `createApi` has some gnarly return types, and i wasn't able to abstract it to the pattern i wanted in a reasonable amount of time.

i have moved the flag and everything inside this api, and will assist with whoever needs to add additional rebasing tokens in future.

**Review by toshiSat — APPROVED** [review]

**Review by 0xApotheosis — APPROVED** [review]
👌

**Pull request merged**